framework/wazuh/agent.py

```diff
@@ -676,55 +680,124 @@ def get_outdated_agents(agent_list=None, offset=0, limit=common.database_limit,
     return result
 
 
-@expose_resources(actions=["agent:upgrade"], resources=["agent:id:{agent_list}"], post_proc_func=None)
-def upgrade_agents(agent_list=None, wpk_repo=None, version=None, force=False, chunk_size=None, use_http=False):
-    """Read upgrade result output from agent.
-
-    :param agent_list: List of agents ID's.
-    :param wpk_repo: URL for WPK download.
-    :param version: Version to upgrade to.
-    :param force: force the update even if it is a downgrade.
-    :param chunk_size: size of each update chunk.
-    :param use_http: False for HTTPS protocol, True for HTTP protocol.
-    :return: Upgrade message.
-    """
-    # We access unique agent_id from list, this may change if and when we decide to add option to upgrade a list of
-    # agents
-    agent_id = agent_list[0]
-
-    return Agent(agent_id).upgrade(wpk_repo=wpk_repo, version=version, force=True if int(force) == 1 else False,
-                                   chunk_size=chunk_size, use_http=use_http)
-
+@expose_resources(actions=["agent:upgrade"], resources=["agent:id:{agent_list}"],
+                  post_proc_kwargs={'exclude_codes': [1703, 1818]})
+def upgrade_agents(agent_list=None, wpk_repo=None, version=None, force=False, use_http=False,
+                   file_path=None, installer=None):
+    """Start the agent upgrade process.
 
-@expose_resources(actions=["agent:upgrade"], resources=["agent:id:{agent_list}"], post_proc_func=None)
-def get_upgrade_result(agent_list=None, timeout=3):
-    """Read upgrade result output from agent.
+    Parameters
+    ----------
+    agent_list : list
+        List of agents ID's.
+    wpk_repo : str
+        URL for WPK download.
+    version : str
+        Version to upgrade to.
+    force : bool
+        force the update even if it is a downgrade.
+    use_http : bool
+        False for HTTPS protocol, True for HTTP protocol.
+    file_path : str
+        Path to the installation file.
+    installer : str
+        Selected installer.
 
-    :param agent_list: List of agents ID's.
-    :param timeout: Maximum time for the call to be considered failed.
-    :return: Upgrade result.
+    Returns
+    -------
+    ID of created tasks
     """
-    # We access unique agent_id from list, this may change if and when we decide to add option to upgrade a list of
-    # agents
-    agent_id = agent_list[0]
+    result = AffectedItemsWazuhResult(all_msg='All upgrade tasks have been created',
+                                      some_msg='Some upgrade tasks have been created',
+                                      none_msg='No upgrade task has been created',
+                                      sort_fields=['task_id'], sort_ascending='True')
+
+    agent_list = list(map(int, agents_padding(result=result, agent_list=agent_list)))
+    if not file_path:
+        wpk_repo = wpk_repo if wpk_repo else common.wpk_repo_url_4_x
+    if version and not version.startswith('v'):
+        version = f'v{version}'
```

Comment:
I would add a validator to the spec so the version always has v in from of the numbers instead of here. The framework should be unaware of such things. Enforce the user at top level.

Reply:
I added a validator in order to  resolve this conversation. After talking to David, we have decided to remove the validator and let the framework add the `v` in order to avoid future 4.0 user problems.